scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1133,47 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForceToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
+    forceBefore(typedPipe).cross(tiny)
+
+  override def flatMap[U](f: T => TraversableOnce[U]): TypedPipe[U] =
+    forceBefore(typedPipe).flatMap(f)
+
+  private def forceBefore(pipe: TypedPipe[T]): TypedPipe[T] = pipe match {
+    case fToDiskPipe: ForceToDiskTypedPipe[_] => fToDiskPipe // can't move something the user did
+    case typedPipeInst: TypedPipeInst[_] =>
+      if (isMaterialized(typedPipeInst.inpipe))
+        typedPipeInst
+      else
+        physicalForceToDisk
+    case _ => physicalForceToDisk
+  }
+
+  private def isMaterialized(pipe: Pipe): Boolean = pipe match {
```

Comment:
ah nice. To confirm, this is indicated getPrevious array being empty right? (and not the `parent` field on the Pipe :-) )

On stepping through using a debugger it seems like the parent field is always null. I see previous seems to have data in most cases (apart from situations like when the Pipe is a CoGroup / HashJoin / .. in those cases, previous is empty but `pipes` is filled up)